In [1]:
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows',100)

# Read a file
Before pulling up the file, I want to choose coltype because FIPS CODE is numerical code, but many programs recoginized as numerical code, so it deletes leading zero automatically.

In [2]:
# Preselect column type as object
coltype = {'SUMLEV':object,
          'STATE':object,
           'COUNTY':object,
          'PLACE':object,
          'COUSUB':object,
          'CONCIT':object,
          'PRIMGEO_FLAG':object}

In [3]:
pop_2018 = pd.read_csv("sub-est2018_all.csv",encoding="ISO-8859-1", dtype= coltype)

In [4]:
pop_2018.head()

SUMLEV STATE COUNTY  PLACE COUSUB CONCIT PRIMGEO_FLAG FUNCSTAT  \
0         40    01    000  00000  00000  00000            0        A   
1        162    01    000  00124  00000  00000            0        A   
2        162    01    000  00460  00000  00000            0        A   
3        162    01    000  00484  00000  00000            0        A   
4        162    01    000  00676  00000  00000            0        A   

              NAME   STNAME CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010  \
0          Alabama  Alabama       4779736            4780138          4785448   
1   Abbeville city  Alabama          2688               2688             2684   
2  Adamsville city  Alabama          4522               4516             4510   
3     Addison town  Alabama           758                754              751   
4       Akron town  Alabama           356                356              355   

   POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  POPESTIMATE2014  \
0          4798834          4815564          4830460          4842481   
1             2677             2629             2612             2595   
2             4496             4474             4453             4432   
3              751              744              743              740   
4              345              345              341              339   

   POPESTIMATE2015  POPESTIMATE2016  POPESTIMATE2017  POPESTIMATE2018  
0          4853160          4864745          4875120          4887871  
1             2588             2573             2566             2563  
2             4406             4373             4342             4325  
3              735              734              728              724  
4              337              332              332              330

In [5]:
# First, I want to find out how many state FIPS code. I know that it should be 51 including DC.
pop_2018['STATE'].nunique()

51

# There are 51 unique state codes, which is right.

In [6]:
pop_2018.groupby(['STATE', 'STNAME','COUNTY','PLACE']).agg(['count'])['NAME'].reset_index()

STATE   STNAME COUNTY  PLACE  count
0        01  Alabama    000  00000      1
1        01  Alabama    000  00124      1
2        01  Alabama    000  00460      1
3        01  Alabama    000  00484      1
4        01  Alabama    000  00676      1
...     ...      ...    ...    ...    ...
46392    56  Wyoming    043  99990      1
46393    56  Wyoming    045  00000      1
46394    56  Wyoming    045  56215      1
46395    56  Wyoming    045  79125      1
46396    56  Wyoming    045  99990      1

[46397 rows x 5 columns]

In [7]:
pop_2018['COUNTY']=='000'

0         True
1         True
2         True
3         True
4         True
         ...  
81431    False
81432    False
81433    False
81434    False
81435    False
Name: COUNTY, Length: 81436, dtype: bool

## Find out how many county FIPS CODE
## Logic is in each state, COUNTY FIPS CODE with non-zero considers as county and place code is 00000, CONSUB is 00000

In [8]:

county_df = pop_2018[(pop_2018['COUNTY']!='000')&(pop_2018['PLACE']=='00000')&(pop_2018['COUSUB']=='00000')]
county_df.reset_index(drop=True)

SUMLEV STATE COUNTY  PLACE COUSUB CONCIT PRIMGEO_FLAG FUNCSTAT  \
0            50    01    001  00000  00000  00000            0        A   
1            50    01    003  00000  00000  00000            0        A   
2            50    01    005  00000  00000  00000            0        A   
3            50    01    007  00000  00000  00000            0        A   
4            50    01    009  00000  00000  00000            0        A   
...         ...   ...    ...    ...    ...    ...          ...      ...   
3137         50    56    037  00000  00000  00000            0        A   
3138         50    56    039  00000  00000  00000            0        A   
3139         50    56    041  00000  00000  00000            0        A   
3140         50    56    043  00000  00000  00000            0        A   
3141         50    56    045  00000  00000  00000            0        A   

                   NAME   STNAME CENSUS2010POP  ESTIMATESBASE2010  \
0        Autauga County  Alabama         54571              54574   
1        Baldwin County  Alabama        182265             182264   
2        Barbour County  Alabama         27457              27457   
3           Bibb County  Alabama         22915              22920   
4         Blount County  Alabama         57322              57321   
...                 ...      ...           ...                ...   
3137  Sweetwater County  Wyoming         43806              43806   
3138       Teton County  Wyoming         21294              21298   
3139       Uinta County  Wyoming         21118              21121   
3140    Washakie County  Wyoming          8533               8528   
3141      Weston County  Wyoming          7208               7208   

      POPESTIMATE2010  POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  \
0               54754            55208            54936            54713   
1              183111           186540           190143           194886   
2               27330            27350            27174            26944   
3               22872            22747            22664            22516   
4               57373            57554            57570            57611   
...               ...              ...              ...              ...   
3137            43574            43981            45005            45162   
3138            21297            21414            21625            22318   
3139            21089            20893            20994            20953   
3140             8530             8448             8408             8414   
3141             7199             7142             7077             7137   

      POPESTIMATE2014  POPESTIMATE2015  POPESTIMATE2016  POPESTIMATE2017  \
0               54876            54838            55242            55443   
1              199189           202995           207712           212619   
2               26758            26294            25819            25158   
3               22541            22562            22576            22555   
4               57521            57522            57517            57827   
...               ...              ...              ...              ...   
3137            44957            44754            44275            43547   
3138            22777            23016            23161            23261   
3139            20827            20770            20691            20456   
3140             8275             8280             8168             8035   
3141             7138             7197             7213             6986   

      POPESTIMATE2018  
0               55601  
1              218022  
2               24881  
3               22400  
4               57840  
...               ...  
3137            43051  
3138            23081  
3139            20299  
3140             7885  
3141             6967  

[3142 rows x 21 columns]

In [9]:
print("There are {} County FIPS CODE in 2018 population data.".format(county_df.shape[0]))

There are 3142 County FIPS CODE in 2018 population data.


## Which meets excel pivot table number.

In [10]:
# After we have out how many county code, I want to find out how many counties are in each state and show top 5 states

county_code = county_df.groupby(['STATE', 'STNAME']).agg(['count'])['NAME'].\
sort_values(by='count', ascending=False).reset_index()

county_code.head()

STATE    STNAME  count
0    48     Texas    254
1    13   Georgia    159
2    51  Virginia    133
3    21  Kentucky    120
4    29  Missouri    115

## I checked google also and number are matched.

## find out how many PLACE FIPS CODE
## Logic is in each state, PLACE FIPS CODE with non-zero considers as city and COUNTY code is 00000,CONCIT is 00000

In [12]:

city_df = pop_2018[(pop_2018['PLACE']!='00000')&(pop_2018['COUNTY']=='000')&(pop_2018['CONCIT']=='00000')]
city_df.reset_index(drop=True)

SUMLEV STATE COUNTY  PLACE COUSUB CONCIT PRIMGEO_FLAG FUNCSTAT  \
0            162    01    000  00124  00000  00000            0        A   
1            162    01    000  00460  00000  00000            0        A   
2            162    01    000  00484  00000  00000            0        A   
3            162    01    000  00676  00000  00000            0        A   
4            162    01    000  00820  00000  00000            0        A   
...          ...   ...    ...    ...    ...    ...          ...      ...   
19490        162    56    000  81300  00000  00000            0        A   
19491        162    56    000  83040  00000  00000            0        A   
19492        162    56    000  84925  00000  00000            0        A   
19493        162    56    000  85015  00000  00000            0        A   
19494        162    56    000  86665  00000  00000            0        A   

                  NAME   STNAME CENSUS2010POP  ESTIMATESBASE2010  \
0       Abbeville city  Alabama          2688               2688   
1      Adamsville city  Alabama          4522               4516   
2         Addison town  Alabama           758                754   
3           Akron town  Alabama           356                356   
4       Alabaster city  Alabama         30352              31091   
...                ...      ...           ...                ...   
19490   Wamsutter town  Wyoming           451                451   
19491   Wheatland town  Wyoming          3627               3627   
19492     Worland city  Wyoming          5487               5487   
19493      Wright town  Wyoming          1807               1807   
19494       Yoder town  Wyoming           151                151   

       POPESTIMATE2010  POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  \
0                 2684             2677             2629             2612   
1                 4510             4496             4474             4453   
2                  751              751              744              743   
3                  355              345              345              341   
4                31186            31346            31653            31946   
...                ...              ...              ...              ...   
19490              450              454              462              488   
19491             3624             3622             3627             3626   
19492             5487             5435             5419             5420   
19493             1808             1808             1857             1851   
19494              154              153              154              158   

       POPESTIMATE2014  POPESTIMATE2015  POPESTIMATE2016  POPESTIMATE2017  \
0                 2595             2588             2573             2566   
1                 4432             4406             4373             4342   
2                  740              735              734              728   
3                  339              337              332              332   
4                32149            32739            32984            33223   
...                ...              ...              ...              ...   
19490              510              502              494              486   
19491             3645             3650             3596             3565   
19492             5330             5334             5263             5173   
19493             1845             1870             1842             1742   
19494              159              159              156              157   

       POPESTIMATE2018  
0                 2563  
1                 4325  
2                  724  
3                  330  
4                33340  
...                ...  
19490              478  
19491             3553  
19492             5075  
19493             1733  
19494              156  

[19495 rows x 21 columns]

In [13]:
print("There are {} PLACE FIPS CODE in 2018 population data.".format(city_df.shape[0]))

There are 19495 PLACE FIPS CODE in 2018 population data.


In [14]:
city_code = city_df.groupby(['STATE', 'STNAME']).agg(['count'])['NAME'].\
sort_values(by='count', ascending=False).reset_index()

city_code.head()

STATE        STNAME  count
0    17      Illinois   1298
1    48         Texas   1220
2    42  Pennsylvania   1013
3    19          Iowa    943
4    29      Missouri    942

In [15]:
# Combine county and city dataset
join_df = pd.merge(county_code, city_code, on=['STATE', 'STNAME'],suffixes=('_County', '_Place'))
join_df

STATE                STNAME  count_County  count_Place
0     48                 Texas           254         1220
1     13               Georgia           159          538
2     51              Virginia           133          228
3     21              Kentucky           120          416
4     29              Missouri           115          942
5     20                Kansas           105          627
6     17              Illinois           102         1298
7     37        North Carolina           100          551
8     19                  Iowa            99          943
9     47             Tennessee            95          345
10    31              Nebraska            93          529
11    18               Indiana            92          567
12    39                  Ohio            88          931
13    27             Minnesota            87          853
14    26              Michigan            83          533
15    28           Mississippi            82          298
16    40              Oklahoma            77          589
17    05              Arkansas            75          501
18    55             Wisconsin            72          601
19    42          Pennsylvania            67         1013
20    01               Alabama            67          461
21    12               Florida            67          413
22    46          South Dakota            66          312
23    08              Colorado            64          271
24    22             Louisiana            64          304
25    36              New York            62          599
26    06            California            58          482
27    30               Montana            56          129
28    54         West Virginia            55          232
29    38          North Dakota            53          357
30    45        South Carolina            46          271
31    16                 Idaho            44          200
32    53            Washington            39          281
33    41                Oregon            36          240
34    35            New Mexico            33          105
35    02                Alaska            29          149
36    49                  Utah            29          247
37    24              Maryland            24          157
38    56               Wyoming            23           99
39    34            New Jersey            21          324
40    32                Nevada            17           19
41    23                 Maine            16           23
42    04               Arizona            15           91
43    50               Vermont            14           42
44    25         Massachusetts            14           54
45    33         New Hampshire            10           13
46    09           Connecticut             8           30
47    44          Rhode Island             5            8
48    15                Hawaii             5            1
49    10              Delaware             3           57
50    11  District of Columbia             1            1

In [16]:
join_df['Total FIPS CODE'] = join_df['count_County']+join_df['count_Place']

In [17]:
final_df = join_df.sort_values(by='Total FIPS CODE', ascending=False).reset_index(drop=True)

In [18]:
final_df

STATE                STNAME  count_County  count_Place  Total FIPS CODE
0     48                 Texas           254         1220             1474
1     17              Illinois           102         1298             1400
2     42          Pennsylvania            67         1013             1080
3     29              Missouri           115          942             1057
4     19                  Iowa            99          943             1042
5     39                  Ohio            88          931             1019
6     27             Minnesota            87          853              940
7     20                Kansas           105          627              732
8     13               Georgia           159          538              697
9     55             Wisconsin            72          601              673
10    40              Oklahoma            77          589              666
11    36              New York            62          599              661
12    18               Indiana            92          567              659
13    37        North Carolina           100          551              651
14    31              Nebraska            93          529              622
15    26              Michigan            83          533              616
16    05              Arkansas            75          501              576
17    06            California            58          482              540
18    21              Kentucky           120          416              536
19    01               Alabama            67          461              528
20    12               Florida            67          413              480
21    47             Tennessee            95          345              440
22    38          North Dakota            53          357              410
23    28           Mississippi            82          298              380
24    46          South Dakota            66          312              378
25    22             Louisiana            64          304              368
26    51              Virginia           133          228              361
27    34            New Jersey            21          324              345
28    08              Colorado            64          271              335
29    53            Washington            39          281              320
30    45        South Carolina            46          271              317
31    54         West Virginia            55          232              287
32    49                  Utah            29          247              276
33    41                Oregon            36          240              276
34    16                 Idaho            44          200              244
35    30               Montana            56          129              185
36    24              Maryland            24          157              181
37    02                Alaska            29          149              178
38    35            New Mexico            33          105              138
39    56               Wyoming            23           99              122
40    04               Arizona            15           91              106
41    25         Massachusetts            14           54               68
42    10              Delaware             3           57               60
43    50               Vermont            14           42               56
44    23                 Maine            16           23               39
45    09           Connecticut             8           30               38
46    32                Nevada            17           19               36
47    33         New Hampshire            10           13               23
48    44          Rhode Island             5            8               13
49    15                Hawaii             5            1                6
50    11  District of Columbia             1            1                2

In [19]:
final_df.to_csv("joindf.csv",index=False)

# Change in FIPS CODE

How many FIPS codes have been changed. (especially in place code level)
For change, I downloaded the data from Geographic Boundary Change Notes in Census.


https://www.census.gov/programs-surveys/geography/technical-documentation/boundary-change-notes.html
I chose the United States to get all the dataset. 
https://www2.census.gov/geo/docs/reference/bndrychange/united-states.txt




Also in this file, before pulling up the file, I want to choose coltype because FIPS CODE is numerical code, but many programs recoginized as numerical code, so it deletes leading zero automatically.

In [20]:
# Preselect column type as object
coltype = {'State':object,
           'Code':object}

In [26]:
changes_fips_code = pd.read_csv("Boundary_change.csv", dtype= coltype)

In [27]:
changes_fips_code

State Area Name (with parent county and code)   Code  \
0       01         Adamsville city (Jefferson-073)  00460   
1       01         Adamsville city (Jefferson-073)  00460   
2       01                    Bullock County (011)    011   
3       01           Aliceville city (Pickens-107)  01228   
4       01             Anniston city (Calhoun-015)  01852   
...    ...                                     ...    ...   
8609    78       Frederiksted town (St. Croix-010)  37900   
8610    78         Grove Place CDP (St. Croix-010)  46000   
8611    78           Red Hook CDP (St. Thomas-030)  65530   
8612    78               Tutu CDP (St. Thomas-030)  78300   
8613    78               Tutu CDP (St. Thomas-030)  78300   

                     Type of Change Effective Date  Effective_Year  \
0     Additional County Subdivision         1/2/01            2001   
1     Additional County Subdivision         1/3/06            2006   
2        County Boundary Correction         1/1/92            1992   
3     Additional County Subdivision        3/10/08            2008   
4              Merger/Consolidation       12/15/00            2000   
...                             ...            ...             ...   
8609         FIPS Class Code Change        4/17/08            2008   
8610                 Deleted Entity         1/1/10            2010   
8611                     New Entity         1/1/10            2010   
8612  Name Change, FIPS Code Change         1/1/10            2010   
8613                     New Entity         1/1/10            2010   

      Effective_Month Effective_Day  \
0                   1             2   
1                   1             3   
2                   1             1   
3                   3            10   
4                  12            15   
...               ...           ...   
8609                4            17   
8610                1             1   
8611                1             1   
8612                1             1   
8613                1             1   

                                  Description of Change  \
0     Annexed into Robbins Crossroads (92745) and We...   
1                   Annexed into Brookside CCD (90405).   
2     Bullock County (011), Fitzpatrick CCD (91098) ...   
3                  Annexed into Carrollton CCD (90495).   
4     Most area of deleted Blue Mountain town (07552...   
...                                                 ...   
8609  FIPS Class Code (CLASSFP) corrected in the MTD...   
8610  Formerly located in Northwest subdistrict (622...   
8611  Located in St. Thomas Island (030), East End s...   
8612  Name changed from Anna's Retreat (05500). FIPS...   
8613  Located in St. Thomas Island (030), East End (...   

                    Source of Change Date Submitted  
0     Boundary and Annexation Survey         5/8/03  
1     Boundary and Annexation Survey         7/2/10  
2     Boundary and Annexation Survey        6/20/01  
3     Boundary and Annexation Survey        7/28/09  
4     Boundary and Annexation Survey         3/5/03  
...                              ...            ...  
8609                   Census Bureau        4/17/08  
8610            Island Area Programs        8/26/11  
8611            Island Area Programs       12/22/10  
8612            Island Area Programs       12/22/10  
8613            Island Area Programs       12/22/10  

[8614 rows x 11 columns]

There are 8,614 cases of change from 1870 - 2010

Then, How many changes happened in a year?

In [28]:
changes_fips_code.groupby('Effective_Year').agg('count')['Code']

Effective_Year
1870       2
1890       1
1891       1
1892       2
1893       4
1912       1
1915       1
1919       1
1956       2
1958       1
1960       1
1963      14
1964       1
1965       3
1966       1
1967       2
1969       1
1970       2
1971       3
1972       3
1973       1
1974       1
1975       1
1976       2
1977       1
1978       1
1979       2
1980       4
1981       5
1982       1
1983       1
1984       4
1985       2
1986       7
1987       5
1988       3
1989       4
1990     127
1991      29
1992      32
1993      14
1994      18
1995      10
1996      28
1997      35
1998      56
1999     186
2000     718
2001     466
2002     402
2003     329
2004     244
2005     468
2006     309
2007     369
2008     402
2009     842
2010    2181
2011     769
2012     412
2013      76
Name: Code, dtype: int64

In [45]:
# Check what will be the biggest cause of change
# I will groupby Type of Change
reason_df = changes_fips_code.groupby('Type of Change')['Code'].agg('count').reset_index().sort_values(by='Code', ascending=False)

In [54]:
#Top 5 Reason
reason_df.rename(columns={"Code": "Number of Change"}, inplace=True)
reason_df.head().reset_index(drop=True)

Type of Change  Number of Change
0  Additional County Subdivision              1125
1  FIPS Code Change, Name Change               824
2  Redistricting/Reapportionment               606
3     County Boundary Correction               574
4                     New Entity               551

In [49]:
# Let's check changes after year 1999 because that is why most changes are occurred.

changes_after_1999= changes_fips_code[changes_fips_code['Effective_Year']> 1998]

In [50]:
reason_df_1999 = changes_after_1999.groupby('Type of Change')['Code'].agg('count').reset_index().sort_values(by='Code', ascending=False)

In [55]:
reason_df_1999.rename(columns={"Code": "Number of Change"}, inplace=True)
reason_df_1999.head().reset_index(drop=True)

Type of Change  Number of Change
0  Additional County Subdivision              1082
1  FIPS Code Change, Name Change               814
2  Redistricting/Reapportionment               594
3     County Boundary Correction               564
4                     New Entity               534

In [56]:
reason_df_1999

Type of Change  Number of Change
1                       Additional County Subdivision              1082
43                      FIPS Code Change, Name Change               814
75                      Redistricting/Reapportionment               594
19                         County Boundary Correction               564
65                                         New Entity               534
26                                     Deleted Entity               524
16                                Boundary Correction               516
33                   Detached From County Subdivision               495
29                                 Description Change               316
21             County Subdivision Boundary Correction               314
58                                        Name Change               288
18                             County Boundary Change               252
37                                   FIPS Code Change               219
0                                   Additional County               200
36                             FIPS Class Code Change               157
20                 County Subdivision Boundary Change               155
15                                    Boundary Change               151
64                             New County Subdivision               125
61                      Name Change, FIPS Code Change               123
50                               Merger/Consolidation               118
68                    Place Affected by Redistricting               106
25                         Deleted County Subdivision                66
32                               Detached From County                62
54                                     NS Code Change                61
11            Became Inactive, FIPS Class Code Change                55
74                           Redistricting Correction                53
35                   Detached from County Subdivision                35
48               Legal/Statistical Description Change                29
12                                 Became Independent                29
22           County Subdivision No Longer Functioning                19
6               Became Active, FIPS Class Code Change                12
23             County Subdivision Relationship Change                11
28                         Deleted county subdivision                10
40           FIPS Code Change, FIPS Class Code Change                 8
72     Place No Longer Active, FIPS Class Code Change                 6
27                   Deleted Entity, Became Dependent                 5
5                                     Area Transposed                 4
47   Functional Status Change, FIPS Class Code Change                 3
31  Description Change, Name Change, FIPS Code Change                 3
9            Became Dependent, FIPS Class Code Change                 3
42                   FIPS Code Change, NS Code Change                 3
45           FIPS Code Change; FIPS Class Code Change                 3
66                             New county subdivision                 2
57                                 NS Code Correction                 2
62  Name Change, FIPS Code Change, NS Code Change,...                 2
73             Reclassification of County Subdivision                 2
70  Place Affected by Redistricting, Additional Co...                 2
38                FIPS Code Change and NS Code Change                 2
76                    Removed from County Subdivision                 2
2   Additional County Subdivision. Deleted County ...                 2
8                                    Became Dependent                 2
39  FIPS Code Change, Description Change, Name Change                 2
30                    Description Change, Name Change                 2
17                                  Class Code Change                 1
3                   Additional County, Deleted County                 1
4      